### PR TITLE
Wandb dashboard

### PR DESCRIPTION
**Pull request by soumik12345**
- [x] Add wandb tables in GAN Monitor
- [ ] Replace ml_collections with wandb configs
- [ ] Save checkpoints every epoch interval
- ~[ ] Integrate sweeps~
- [ ] Save checkpoints as artifacts

### CODE REVIEW

**Comment by soumik12345** [comment]
@sayakpaul Sample run link with tables: https://wandb.ai/tf2_gans/GauGAN/runs/ex7z0ia0

**Comment by sayakpaul** [comment]
Why integrating sweeps is being considered here?

**Comment by soumik12345** [comment]
> Why integrating sweeps is being considered here?

Do you think we can perform a sweep just as an experiment to see if hyperparameter tuning adds any value for the larger datasets? If not, we can strike it off.

**Comment by sayakpaul** [comment]
I think we can do that as a separate project to keep the current goals up and moving. Demonstrating good results with Coco-Stuff 10k sounds fare more interesting to me. This is also because the GauGAN hyperparameters are quite well-studied. But I am definitely welcoming of the sweeps projects, though just not immediately.

File: gaugan/callbacks.py [hunk 1]
```diff
@@ -31,32 +34,38 @@ def infer(self):
...
-                fig.savefig(os.path.join(self.plot_save_dir, f"{epoch}.png"))
+                fig = plot_results(
+                    [semantic_input[i], ground_truth[i], generated_images[i]],
+                    self.columns,
+                    figure_size=(18, 18),
+                )
+                if (self.plot_save_dir is None) and (not self.use_wandb):
+                    plt.show()
+                elif self.use_wandb:
+                    wandb.log({f"validation_images_{epoch + 1}_{i}": fig})
```
Comment (sayakpaul) [inline gaugan/callbacks.py:69]: No need to use this if we are using `wandb.Table` I suppose?

**Review by sayakpaul — APPROVED** [review]

**Comment by sayakpaul** [comment]
> @sayakpaul Sample run link with tables: https://wandb.ai/tf2_gans/GauGAN/runs/ex7z0ia0

The tables look pretty nice.

File: gaugan/callbacks.py (hunk 1 above)
Comment (soumik12345) [reply gaugan/callbacks.py]: Done

File: gaugan/callbacks.py [hunk 2]
```diff
@@ -31,32 +34,50 @@ def infer(self):
...
+            )
+        wandb.log({f"GANMonitor Epoch {epoch}": wandb_table})
+
     def on_epoch_end(self, epoch, logs=None):
         if epoch == 0 or (epoch + 1) % self.epoch_interval == 0:
             generated_images = self.infer()
 
-            grid_row = min(generated_images.shape[0], 3)
-            fig, axarr = plt.subplots(4, 3, figsize=(18, grid_row * 6))
+            semantic_input = (self.val_images[0] + 1) / 2
```
Comment (scottire) [inline gaugan/callbacks.py:51]: maybe this could be a function

File: gaugan/callbacks.py [hunk 3]
```diff
@@ -31,32 +34,50 @@ def infer(self):
         )
         return self.model.predict([latent_vector, self.val_images[2]])
 
+    def log_to_tables(self, epoch, semantic_input, ground_truth, generated_images):
+        wandb_table = wandb.Table(columns=self.columns)
+        for i in range(4):
```
Comment (scottire) [inline gaugan/callbacks.py:39]: this 4 number could be explained or made a global variable. It's currently a bit of a magic number

**Review by scottire — APPROVED** [review]

**Comment by sayakpaul** [comment]
@soumik12345, I am taking care of Scott's comments. Please resolve the conflicts afterward.

File: gaugan/callbacks.py (hunk 2 above)
Comment (sayakpaul) [reply gaugan/callbacks.py:51]: @scottire since it's just a one-liner preprocessing op, I think it's fine to leave it as it is.

**Pull request merged**